### PR TITLE
PostgreSQL 10 support

### PR DESCRIPTION
**Pull request by simPod**
When using Postgres 10, DBAL throws

```
An exception occurred while executing 'SELECT min_value, increment_by FROM "schemaname"."table_id_seq"':

  SQLSTATE[42703]: Undefined column: 7 ERROR:  column "min_value" does not exist
  LINE 1: SELECT min_value, increment_by FROM "schemaname"."table_id_se...
```

This fixes it as PSQL10 reads sequences definitions from `pg_sequences`. https://www.postgresql.org/docs/10/static/view-pg-sequences.html

### CODE REVIEW

File: lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php
Comment (lcobucci) [inline lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: This is breaking the test suite, can you please check that?

Comment (lcobucci) [reply lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: Also [Postgres 10 yet is not supported in Travis](https://docs.travis-ci.com/user/database-setup/#Using-a-different-PostgreSQL-Version), so there's no way to 100% guarantee the compatibility at this moment.

**Review by lcobucci — CHANGES_REQUESTED** [review]
Can you also add new tests that covers that for PGSQL 10?

**Comment by Majkl578** [comment]
Duplicate of #2868. Waiting for Travis, see https://github.com/travis-ci/travis-ci/issues/8537.

**Comment by simPod** [comment]
@lcobucci It's done. I believe the test was not working properly before (correct me if I'm wrong)

**Comment by simPod** [comment]
There was a case when null was returned for $data and this case was ignored. https://github.com/doctrine/dbal/pull/2893/files#diff-e4bd68e20fcb55598e13e0831ba281abR63 I can add null case it needed.

File: lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php
Comment (Majkl578) [inline lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: I think this should rather be based on Platform check, there will sure be separate platform for Postgres 10. Anyway I think it should all be done in a single PR, not fixing random pieces randomly... :/

Comment (lcobucci) [reply lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: I agree, also exposing `Connection#getServerVersion()` a no-go IMO

**Comment by simPod** [comment]
Tried to refactor. I see this project's internals for the first time tho but IMO it's good start for PSQL10. I can either continue improving this or we can close.

**Comment by PillowPillow** [comment]
Any news about this PR?

**Comment by iquad** [comment]
@PillowPillow Travis still working to implement Postgresql 10. Until that time, we have to wait.

**Comment by SCIF** [comment]
Just ran functional tests on Postgres10 localy and all good. :+1:

**Comment by lcobucci** [comment]
@SCIF it's really good to hear that let's just hope that travis adds support to it soon

**Comment by Majkl578** [comment]
I got tired of infinitely waiting for Travis, so here is externally installed Postgres 10: #2946. Feel free to pick that commit and build a working compat PR on top. 😎

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (Majkl578) [inline lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: DBAL requires PHP 7.1, please add parameter and return types where applicable.

File: lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL100Keywords.php
Comment (Majkl578) [inline lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL100Keywords.php]: No license headers please, we've already removed them in develop.

File: tests/Doctrine/Tests/DBAL/Driver/AbstractPostgreSQLDriverTest.php
Comment (Majkl578) [inline tests/Doctrine/Tests/DBAL/Driver/AbstractPostgreSQLDriverTest.php]: Please use ::class.

File: tests/Doctrine/Tests/DBAL/Schema/PostgreSQLSchemaManagerTest.php
Comment (Majkl578) [inline tests/Doctrine/Tests/DBAL/Schema/PostgreSQLSchemaManagerTest.php]: Please use ::class.

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: It should probably be renamed to `getSequenceMetadataSQL`, `data` is not explicit.

**Review by Majkl578 — CHANGES_REQUESTED** [review]

**Comment by Majkl578** [comment]
Here's a diff of keywords in [9.6](https://www.postgresql.org/docs/9.6/static/sql-keywords-appendix.html) and [10](https://www.postgresql.org/docs/10/static/sql-keywords-appendix.html): https://www.diffchecker.com/fJ0YkHgw
There are no new reserved keywords, but some of the existing are not reserved anymore. But I'm not a DBAL guy so I can't say whether they are supposed to be removed in PostgreSQL100Keywords or not.

**Comment by simPod** [comment]
They were not reserved before tho.

File: lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php]: There shouldn't be a semantic difference between this method and its parents from `AbstractPlatform` and `PostgreSqlPlatform`. Otherwise, the caller which depends on the abstract name will not know whether it should pass raw or non-raw sequence name.

File: lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: Is it really necessary to `fetchAll()` and then use only `$data[0]`? Doing `fetch()` and using the row directly would make more sense to me.

File: lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php]: Given that in PostgreSQL 10, the sequence metadata is available via a view, would it make sense to rework `getListSequencesSQL()` instead? With this approach, the client will not have to execute the SQL from  `getSequenceMetadataSQL()` for each sequence, and the method itself will not need to be introduced on the abstract platform level.

**Comment by iquad** [comment]
@simPod are you going to fix/apply or need help ? We're waiting to deploy new product to  production for 2 months :P I can help if you wish.

File: lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php
Comment (simPod) [reply lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: There's no method `fetch()` in connection. In my opinion, this is ok as it fetches one row

File: lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php
Comment (simPod) [reply lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php]: Not sure it's a good idea. `getListSequencesSQL()` lists sequences. It would need to add also `increment_by` and `min_value` columns. That would change the method's functionality. This method is called before in `listSequences()` already.

**Comment by simPod** [comment]
@iquad you can already use this fork so you don't have to wait :)

**Comment by iquad** [comment]
@simPod yeah, but need doctrine to merge them, server tests, 30+ development machine etc. Can't handle them one by one :/

**Comment by simPod** [comment]
@iquad personally I think it's done but waiting for approvals or some other dbal guys reviews

File: lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php
Comment (Majkl578) [reply lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: There are more fetch*() methods though, fetchArray() is the one you need.

File: lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php
Comment (Majkl578) [inline lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php]: Please address CI failure - spaces around dots.

**Review by Majkl578 — CHANGES_REQUESTED** [review]

File: lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php
Comment (simPod) [reply lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: The data then looks like  
image. 
I personally think the current state is more verbose as I have to reference array values by column names and not by numeric indeces.

File: lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php]: Look at the `OraclePlatform`. Its `getListSequencesSQL()` returns list and metadata. The only reason it doesn't do the same for PostgreSQL is that prior to version 10 there was no view to fetch the metadata from.

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: It doesn't belong here. Please implement `getListSequencesSQL()` for the new platform so that it returns sequence names and metadata same as it's done in the `OraclePlatform`. The only platform which needs to be able to provide metadata for one individual sequence is PostgreSQL < 10 because it doesn't have sequence metadata available via a view.

**Review by morozov — CHANGES_REQUESTED** [review]

**Comment by simPod** [comment]
Any idea how to run that failing test? https://travis-ci.org/doctrine/dbal/jobs/320454235 Have not found any way how to reproduce that locally :/

File: lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: @simPod is this the reason of the test failures?

Comment (simPod) [reply lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: @morozov you're right, thanks! I failed to run those localy but it passes now

File: lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL100Keywords.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL100Keywords.php]: As of https://github.com/doctrine/coding-standard/pull/9, `@since` will be not allowed. The `@link` above doesn't bring much value as well IMO.

File: lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: You can use `Connection::fetchAssoc()` to fetch one row and then `array_merge($sequence, $data)` to not confuse the keys.

Comment (simPod) [reply lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: That would require rework of other tests https://travis-ci.org/doctrine/dbal/jobs/320775111 However, when I launch the testsuite locally, they all pass.

Comment (morozov) [reply lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: Which exactly failures do you think are caused by this?

Comment (simPod) [reply lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: My bad, found it, fixed it

File: lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/PostgreSQL100Platform.php]: Please remove annotations.

File: tests/Doctrine/Tests/DBAL/Schema/PostgreSQLSchemaManagerTest.php
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Schema/PostgreSQLSchemaManagerTest.php]: This test is hard to maintain and not really useful:

1. It used to pass on PostgreSQL 10 before the issue was fixed.
2. It started failing and had to be reworked after the issue was fixed.
3. It enforces implementation details which are not necessarily correct (`fetchAssoc` vs. `fetchAll`).

Let's replace this test with a functional one in `SchemaManagerFunctionalTestCase`. The test should create a sequence using the `AbstractSchemaManager` manager API, and then list the sequences to see if it's created and introspected correctly. This test will not depend on the implementation details of a given platform and will prove that the functionality is portable between the platforms.

**Review by morozov — COMMENTED** [review]
The code looks good. Now it's time to take care of the test :-)

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (Majkl578) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: This seems to error on Oracle now:
```
1) Doctrine\Tests\DBAL\Functional\Schema\OracleSchemaManagerTest::testCreateAndListSequences 
Exception: [PHPUnit\Framework\Error\Notice] Undefined index: sequence_1 
```
https://continuousphp.com/git-hub/doctrine/dbal/build/c3aeacf9-fd40-45c2-8304-e93fc34a5bee/output

Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Yup, have to setup oracle here to reproduce.

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: @simPod please make sure all sequence attributes are covered by the test, not only the name.

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Most likely, it’s upper case on Oracle.

Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Indeed, it is

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Can you review now please? Thanks

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Oracle is not the only supported platform which returns upper-cased column names. DB2 is the other one but it looks DBAL doesn't support sequences in DB2. In any event, checking the platform name is not the best approach.

I can think of two solutions:

1. Figure out why the rest of the tests in this class don't fail on Oracle without using the portability layer. For instance, `testListTableColumns()` expects column names to be lower-cased, however, they are upper-cased in Oracle. Must be implemented in:
https://github.com/doctrine/dbal/blob/22844d3c0d52e44c312ad60571b3d9e8da662c9b/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L824
2. Use `Portability\Connection` with `PORTABILITY_FIX_CASE` and `fetch_case = CASE_LOWER` when instantiating the schema manager. This is hor a real production code would deal with the case discrepancy.

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: It's better to use different values for different parameters to make sure they are not confused during deployment or introspection.

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Not sure the assertions are intentionally this verbose, but the same could be checked in a more compact way:
```php
$sequence1 = new Sequence('sequence_1', 1, 2);
$sequence2 = new Sequence('sequence_2', 3, 4);

// create the sequences

self::assertArraySubset([
    $sequence1,
    $sequence2,
], $this->_sm->listSequences());

Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: I'm fan of verbosity but can remake if you think it better suits code style of this project. Should I proceed then? :)

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Depending on what's the purpose of this verbosity. The approach I suggested is more state-based. It only tests that the collection returned by `listSequences()` contains the objects which were previously created. Isn't it what it's supposed to do?

The current approach also tests the public API of the `Sequence` class. For instance, if there was some logic in `getInitialValue` and `getAllocationSize()`, the test would have to take this logic into account as well which would make it less stable.

There are no strict coding guidelines in this regard, so it's up to you how you want to structure the test.

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: `testListTableColumns` also converts table name to lowercase `self::assertEquals('test', strtolower($columns['test']->getname()));`

Can do it this way then. In https://github.com/doctrine/dbal/blob/22844d3c0d52e44c312ad60571b3d9e8da662c9b/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L824 it just creates an array with lowercased table names as keys.

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: It would be a breaking change. Doing `strtolower()` in a test would mean that the production code should do the same by design which it not true, it's a bug.

Let's skip the test for Oracle for now and fix the schema manager in 3.0.

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: I agree. However, assertArraySubset asserts by keys and there are other Sequences created in db as well.

image

image

I reworked it a bit

Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: It's failing on Psql 9.5. I'd have to either sort it by name or revert to the previous verbose version.

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Eh… I didn't know `assertArraySubset()` is strict about numeric keys. The last approach could fail on any DB because the order of sequences is not enforced.

The problem is not verbosity itself, the problem is also that having the logic like `array_map()` or `array_filter()` outside of assertions which makes the test less readable and less clear about which exactly case it covers.

How about this:
```php
$list = $sm->listSequences();
self::assertContains($sequence1, $list);
self::assertContains($sequence2, $list);
```

Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: That wouldn't work as those are different objects.

I guess in the end this is the simplier solution:

```php
$this->assertSame($sequence1Name, $actualSequence1->getName());
$this->assertEquals($sequence1AllocationSize, $actualSequence1->getAllocationSize());
$this->assertEquals($sequence1InitialValue, $actualSequence1->getInitialValue());
```

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Okay, let's leave it like this.

Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Kk, it's there

**Review by morozov — APPROVED** [review]

File: lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL100Keywords.php
Comment (Majkl578) [inline lib/Doctrine/DBAL/Platforms/Keywords/PostgreSQL100Keywords.php]: typo: `10.0`

File: lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php
Comment (Majkl578) [inline lib/Doctrine/DBAL/Schema/PostgreSqlSchemaManager.php]: Is array_merge needed here? Wouldn't `+` _just_ work since it's _assoc_?

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (Majkl578) [inline tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Please call all asserts statically.

File: tests/Doctrine/Tests/DBAL/Driver/AbstractPostgreSQLDriverTest.php
Comment (Majkl578) [inline tests/Doctrine/Tests/DBAL/Driver/AbstractPostgreSQLDriverTest.php]: This one should be above Doctrine\DBAL\Platforms\PostgreSqlPlatform.

**Review by Majkl578 — CHANGES_REQUESTED** [review]
Just last few nitpicks, otherwise LGTM. 👍

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Isn't that deprecated?

**Comment by simPod** [comment]
@morozov thanks for help!

File: tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php
Comment (Majkl578) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: What do you mean? All assert methods are declared static: https://github.com/sebastianbergmann/phpunit/blob/master/src/Framework/Assert.php

Comment (simPod) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: Ah, not deprecated but it's there because of something oldhttps://github.com/sebastianbergmann/phpunit/issues/1914#issuecomment-148407321. 

Both can be used tho. I'm used to call it via non-static way. 

I guess it needs to be reworked to static then as the whole project uses static calls. Shall I?

Comment (Majkl578) [reply tests/Doctrine/Tests/DBAL/Functional/Schema/SchemaManagerFunctionalTestCase.php]: And bad design in PHPUnit apparently, since non-static TestCase extends static Assert class. :/
Since it's declared as static, it was decided to use static calls in tests for assert methods, otherwise it's different context (and personally I think PHP should emit a warning, just like static->instance call).

**Review by Majkl578 — APPROVED** [review]

**Comment by simPod** [comment]
@lcobucci is there anything else?

**Review by Ocramius — APPROVED** [review]
LGTM :ship:

**Comment by sergiors** [comment]
OMG!

**Comment by iquad** [comment]
thank you so much, all of you!

**Comment by anonim1133** [comment]
Is there any date for 2.7 so I could use postgres 10? Or is there a way to patch 2.6 with this commit?

**Comment by Ocramius** [comment]
No date yet. You can already use 2.6 with PostgreSQL 10, but the schema tools won't be working correctly yet.

**Comment by nicholasruunu** [comment]
Still have this issue with `v2.9.2` and postgres `10.10`

**Comment by exebece** [comment]
I think this issue can easily be solved using proper doctrine connection string specifying the version of postgres like: DATABASE_URL=postgres://user@127.0.0.1:5432/db_name?charset=utf8&serverVersion=10

the serverVersion part helped me solving all issues with Doctrine & Postgres. Try it for free ;)

**Comment by nicholasruunu** [comment]
@exebece Yeah, my bad, updated the doctrine configuration from `9.x` to `10` and it works just fine now. Thanks.

**Comment by XanderKon** [comment]
@exebece Thanks! Spent two hours and your comment save me from another time wasting.

**Pull request merged**
